google/cloud/storage/internal/object_requests.cc

```diff
@@ -330,20 +330,81 @@ std::ostream& operator<<(std::ostream& os, ResumableUploadRequest const& r) {
   os << "ResumableUploadRequest={bucket_name=" << r.bucket_name()
      << ", object_name=" << r.object_name();
   r.DumpOptions(os, ", ");
-  return os << ", payload=" << r.json_payload() << "}";
+  return os << "}";
 }
 
-std::ostream& operator<<(std::ostream& os, UploadChunkRequest const& r) {
-  os << "ResumableUploadRequest={upload_session_url=" << r.upload_session_url()
-     << ", range=" << r.range_begin() << "-" << r.range_end() << "/";
-  if (r.final_chunk()) {
-    os << r.range_end();
+std::string UploadChunkRequest::RangeHeader() const {
+  std::ostringstream os;
+  os << "Content-Range: bytes " << range_begin() << "-";
+  if (payload().empty()) {
+    // This typically happens when the sender realizes too late that the
+    // previous chunk was really the last chunk (e.g. the file is exactly a
+    // multiple of the quantum, reading the last chunk does not detect the EOF),
+    // the formatting of the range is special in this case.
+    os << range_begin();
   } else {
-    os << "*";
+    os << range_begin() + payload().size() - 1;
   }
+  if (source_size() == 0) {
+    os << "/*";
+  } else {
+    os << "/" << source_size();
+  }
+  return std::move(os).str();
+}
+
+std::ostream& operator<<(std::ostream& os, UploadChunkRequest const& r) {
+  os << "UploadChunkRequest={upload_session_url=" << r.upload_session_url()
+     << ", range=<" << r.RangeHeader() << ">";
   r.DumpOptions(os, ", ");
   return os << ", payload="
-            << BinaryDataAsDebugString(r.payload().data(), r.payload().size())
+            << BinaryDataAsDebugString(r.payload().data(), r.payload().size(),
+                                       128)
+            << "}";
+}
+
+std::ostream& operator<<(std::ostream& os,
+                         QueryResumableUploadRequest const& r) {
+  os << "QueryResumableUploadRequest={upload_session_url="
+     << r.upload_session_url();
+  r.DumpOptions(os, ", ");
+  return os << "}";
+}
+
+ResumableUploadResponse ResumableUploadResponse::FromHttpResponse(
+    HttpResponse&& response) {
+  ResumableUploadResponse result;
+  result.last_committed_byte = 0;
+  result.payload = std::move(response.payload);
+  if (response.headers.find("location") != response.headers.end()) {
+    result.upload_session_url = response.headers.find("location")->second;
+  }
+  if (response.headers.find("range") != response.headers.end()) {
```

Comment:
This was a bit hard to follow until I pulled up the docs at https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload. Might be worth adding a comment here, e.g. `We expect a range header with its value formatted like "bytes=0-<number>", where <number> is the number of bytes received so far."  This explains that we always expect 0 as the first number in that range.

Actually, that being said... I think we could simplify lines 385 through 393 -- can we just search for the substring "bytes=0-" and return `result` if that's not found?

Reply:
Thanks!  Fixed I think.